### PR TITLE
Harden nightly eval reporting (Fixes #2605)

### LINKED ISSUE #2605: Fix nightly eval configuration and result reporting
## Summary

The dedicated Evals Nightly workflow has never completed successfully. The recent scheduled run 29719172016 failed all three matrix samples before making a model request because the workflow does not export the required model and StepFun base URL configuration.

A separate report-path defect writes Vitest JSON output outside the uploaded artifact directory. The aggregate job then treats missing reports as success, which can hide broken result collection.

The broader Nightly Tests workflow runs the same behavioral eval with the complete StepFun configuration and passed at the same source revision. This demonstrates that the existing repository provider variables and StepFun credential can work; the primary defects are workflow configuration drift and reporting behavior.

## Evidence

Linked run: https://github.com/vybestack/llxprt-code/actions/runs/29719172016/job/88278298073

The dedicated workflow exports the API key and provider but omits:

- LLXPRT_DEFAULT_MODEL
- OPENAI_BASE_URL

The job fails in packages/test-utils/src/cli-args.ts with:

    LLXPRT_DEFAULT_MODEL environment variable is required but not set

Vitest runs with evals as its root while evals/vitest.config.ts specifies evals/logs/report.json. This resolves to evals/evals/logs/report.json, but the workflow uploads only evals/logs.

scripts/aggregate_evals.js prints "No reports found." and exits successfully when no report is available. In the linked run, each uploaded artifact contained only an empty tool-call log and no report.json, yet Aggregate Results succeeded.

## Current eval coverage

There is currently one real-model behavioral eval in evals/save_memory.eval.ts. It prompts the model to remember and recall that the user's favorite color is blue and requires use of the save_memory tool.

The current validation requires a save_memory tool call and nonempty output, but absence of the expected word "blue" only warns instead of failing. The suite therefore does not fully enforce the behavior described by the prompt.

## Proposed fixes

1. Share one canonical behavioral-eval environment/setup between the dedicated statistical workflow and the broader nightly workflow so provider configuration cannot drift again.
2. Export the complete known-working configuration:
   - OPENAI_API_KEY from the secret named by KEY_VAR_NAME
   - OPENAI_BASE_URL from the repository variable
   - LLXPRT_DEFAULT_MODEL from the repository variable
   - LLXPRT_DEFAULT_PROVIDER from the repository variable
   - LLXPRT_AUTH_TYPE=provider
   - LLXPRT_FORCE_FILE_STORAGE=true
   - RUN_EVALS=1
3. Add a preflight that fails clearly when any required provider setting is empty without exposing secret values.
4. Build explicitly before running the eval suite.
5. Correct the Vitest JSON output path so report.json is written under evals/logs and uploaded with the tool-call logs.
6. Make artifact upload fail when expected files are absent.
7. Make aggregation fail when it finds no reports or no usable assertions instead of producing a false-success job.
8. Add behavioral tests for aggregation failures and result processing, including absent reports, malformed reports, empty assertion data, and valid pass/fail input.
9. Strengthen the save_memory eval so missing expected semantic content fails rather than merely warning, while preserving real-model behavioral validation.
10. Align historical aggregation depth with artifact retention so requested history can actually be downloaded.
11. Correct eval documentation that claims ALWAYS_PASSES evals run on pull requests unless that check is wired into PR CI as part of this work.

## Test-first implementation plan

1. Add failing tests that exercise aggregate_evals.js through its observable process behavior and prove that missing or unusable reports produce a nonzero exit while valid reports produce the expected summary.
2. Add a failing configuration/path test that proves eval reports are emitted beneath the artifact directory when Vitest uses evals as its root.
3. Add a failing workflow validation test using the repository's established workflow-testing conventions, or an equivalent deterministic configuration test, proving all required environment values are wired from repository properties.
4. Add a failing behavioral test for expected eval content, then update the validation call site/API so semantically incorrect output cannot pass.
5. Implement the smallest production and workflow changes needed to make each test pass, then refactor shared nightly setup to remove duplicated provider wiring.
6. Run focused tests during red/green development.
7. Run the complete repository verification suite:
   - npm run test
   - npm run lint
   - npm run typecheck
   - npm run format
   - npm run build
   - bun scripts/start.ts --profile-load ollamakimi "write me a haiku and nothing else"
8. After merge, manually dispatch Evals Nightly on main and confirm:
   - all three matrix samples reach the provider and execute the eval;
   - every artifact contains the tool log and report.json;
   - the aggregate summary contains real pass-rate data;
   - absent reports make aggregation fail;
   - the final workflow conclusion reflects actual eval outcomes.

## Acceptance criteria

- The dedicated Evals Nightly workflow receives the complete StepFun/OpenAI-compatible provider configuration without exposing credentials.
- Provider configuration is defined through a shared mechanism rather than duplicated, drift-prone workflow blocks.
- All matrix samples execute the real behavioral eval instead of failing during local configuration validation.
- JSON reports are uploaded in each matrix artifact.
- Aggregation fails loudly when result data is absent or unusable.
- Aggregation tests cover success and failure behavior without network or model mocks standing in for the code under test.
- The save_memory eval cannot pass when required expected content is missing.
- Documentation accurately describes when eval policies run.
- The complete local verification suite passes.

### PR DESCRIPTION
**Pull request by acoliver**
## TLDR

Makes nightly behavioral evals fail closed and remain diagnosable when execution or artifact generation breaks. Both nightly entry points now use one reusable workflow, save available diagnostics on every attempt, and feed strict current and historical reports into a deterministic aggregator.

## Dive Deeper

- Introduces a reusable behavioral-eval workflow shared by the dedicated eval schedule and the broad nightly workflow.
- Runs the dedicated eval workflow as a three-attempt matrix with stable artifact identities, bounded jobs, least-privilege permissions, and retained diagnostic logs even when the required Vitest report is absent.
- Makes the save-memory eval deterministic: the tool call must persist the canonical fact and the model must return the exact requested answer.
- Adds strict TypeScript compilation and type-aware ESLint coverage for eval source.
- Decomposes aggregation into schema, artifact-cardinality, and historical-run modules.
- Rejects missing, duplicate, unexpected, symlinked, malformed, contradictory, or unusable current artifacts instead of silently publishing partial data.
- Retrieves historical workflow runs through the paginated GitHub API, validates external run metadata, enforces the artifact retention window, deduplicates runs, and omits malformed historical runs atomically.
- Keeps model-output diagnostics privacy-safe by reporting metadata rather than raw output.
- Documents the actual nightly execution and reporting contract.

## Reviewer Test Plan

1. Run `npm run test:scripts` and confirm the aggregation, workflow, exact save-memory, and eval static-compliance suites pass.
2. Run `npm run test --workspace @vybestack/llxprt-code-test-utils` and confirm privacy and regex-state tests pass.
3. Run `npm run lint`, `npm run typecheck`, `npm run format`, and `npm run build`.
4. Run `actionlint -ignore 'SC2129' .github/workflows/_evals-run.yml .github/workflows/evals-nightly.yml .github/workflows/nightly.yml`.
5. Exercise aggregation against a complete `eval-logs-1..3` tree, then remove, duplicate, or corrupt one report and verify aggregation exits nonzero with a branch-specific diagnostic.
6. Inspect a nightly run and confirm all attempts upload stable artifacts, missing reports fail the preparation step while preserving available logs, and aggregation publishes a summary only for complete current input.

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ❓  | CI  |
| npx      | ✅  | ❓  | CI  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

Local macOS verification completed:

- `npm run test` — passed
- `npm run test:scripts` — 2,061 passed, 8 skipped
- test-utils workspace — 25 passed
- `npm run lint` — passed
- `npm run typecheck` — passed
- `npm run format` — passed
- `npm run build` — passed
- Ollama profile smoke — passed with a three-line haiku
- changed workflow actionlint checks — passed
- `git diff --check` — passed

## Linked issues / bugs

Fixes #2605

## Summary by CodeRabbit

* **New Features**
  * Added reusable behavioral evaluation workflows with configurable run attempts and artifact collection.
  * Added historical evaluation aggregation with retention filtering and pass-rate reporting.

* **Bug Fixes**
  * Improved validation for missing, malformed, duplicate, or incomplete evaluation reports.
  * Prevented repeated regular-expression checks from producing inconsistent results.

* **Documentation**
  * Clarified evaluation authoring, execution, reporting, and nightly CI behavior.

* **Tests**
  * Expanded coverage for workflow execution, report validation, aggregation, and deterministic assertions.

### CODE REVIEW

File: .github/workflows/_evals-run.yml [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+          # (eval crashed before the JSON reporter ran). When report.json is
+          # absent, a diagnostic marker is written so post-mortem debugging is
+          # possible, and the job still fails via a later step that checks the
+          # report_present output.
+          rm -rf eval-artifact
+          mkdir -p eval-artifact
+          cp -a evals/logs/. eval-artifact/ 2>/dev/null || true
+          if [ ! -f evals/logs/report.json ]; then
+            echo 'evals/logs/report.json was not produced. The JSON reporter output path is misconfigured or the eval run failed before reporting.' > eval-artifact/REPORT_MISSING.txt
+            echo "::error::evals/logs/report.json was not produced; the JSON reporter output path is misconfigured or the eval run failed before reporting."
```
[the remark this answers is not in the document]
Comment (acoliver) [reply .github/workflows/_evals-run.yml]: Addressed in c447d6f59: reflowed the secret description with YAML folding and split the missing-report diagnostic into shell continuations while preserving behavior. Focused workflow tests, actionlint, ESLint, Prettier, and whitespace checks pass.

**Comment by acoliver** [comment]
CodeRabbit feedback disposition:

- Workflow line-length thread: addressed in c447d6f59 by folding the secret description and splitting the shell diagnostics. Replied inline and resolved the thread.
- Historical isolation test naming/comment finding: addressed in c447d6f59 by renaming the test to match its direct behavior and removing stale contradictory narration.
- Save-memory helper import duplication finding: addressed in c447d6f59 by loading each helper once in beforeAll and reusing suite-scoped functions.
- Docstring coverage warning: evaluated but not expanded into broad test-file documentation work. The production aggregation modules already document their public contracts, and the warning is a repository-independent percentage heuristic over an intentionally test-heavy change rather than a demonstrated correctness or maintainability defect.

Verification for these changes: 64 focused tests passed; ESLint, Prettier, actionlint, and git diff --check passed.

File: packages/test-utils/src/util.ts [hunk 2]
```diff
@@ -154,7 +157,7 @@ export function validateModelOutput(
       'This is not ideal but not a test failure.\n' +
       'The tool was called successfully, which is the main requirement.\n' +
       `Expected content: ${String(expectedContent)}\n` +
-      `Actual output: ${result}`;
+      `Actual output length: ${result.length}`;
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/test-utils/src/util.ts:160]: Retained the privacy-safe diagnostic intentionally. Model output can contain credentials, private prompts, or other sensitive content; logging raw or truncated output would create an avoidable disclosure path. Existing tests verify that warnings report safe metadata without leaking model content.

File: eslint.config.js [hunk 3]
```diff
@@ -610,6 +609,85 @@ export default tseslint.config(
...
+  // await-thenable, return-await) ARE enabled because the eval tsconfig
+  // provides the type program (`project: 'evals/tsconfig.json'`). The eval
+  // source is async-heavy (evalTest runs model calls and tool assertions), so
+  // these rules catch real bugs (unhandled rejections, async functions passed
+  // as non-async callbacks).
+  //
+  // Deliberately NOT applied to evals (documented rationale):
+  //  - strict-boolean-expressions: the eval tsconfig is a standalone noEmit
+  //    project without the full app graph; this rule produces spurious
+  //    cross-package type errors. The other type-aware rules are stable.
```
[the remark this answers is not in the document]
Comment (acoliver) [reply eslint.config.js]: Addressed. The eval ESLint block now enables strict-boolean-expressions with every permissive option disabled. The three resulting eval conditions were made explicit, and the compliance test now protects this rule from regression.

File: .github/workflows/_evals-run.yml [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    timeout-minutes: 30
+    strategy:
+      fail-fast: false
+      matrix:
+        run_attempt: ${{ fromJSON(inputs.run-attempts) }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # ratchet:actions/checkout@v5
+        with:
+          persist-credentials: false
```
[the remark this answers is not in the document]
Comment (acoliver) [reply .github/workflows/_evals-run.yml:34]: No change needed. git rev-parse --short HEAD only needs the checked-out HEAD commit, which is present in a depth-1 checkout; it does not require full history. Keeping the default checkout depth avoids unnecessary transfer.

File: .github/workflows/_evals-run.yml [hunk 5]
```diff
@@ -0,0 +1,142 @@
+name: Reusable Behavioral Evals
+
+on:
+  workflow_call:
+    inputs:
+      run-attempts:
+        description: 'JSON array of matrix run attempt identifiers'
+        required: false
+        type: string
+        default: '[1, 2, 3]'
```
[the remark this answers is not in the document]
Comment (acoliver) [reply .github/workflows/_evals-run.yml:10]: No change needed. This reusable workflow is called only by fixed, repository-controlled workflow definitions; no untrusted event data is forwarded into run-attempts. Matrix expansion also occurs before any job step could validate the value, so a later validation step would not mitigate the described condition.

File: .github/workflows/_evals-run.yml [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+          if [ -z "${LLXPRT_DEFAULT_MODEL:-}" ]; then
+            echo "::error::LLXPRT_DEFAULT_MODEL repository variable is required for real-model evals"
+            missing=1
+          fi
+          if [ -z "${LLXPRT_DEFAULT_PROVIDER:-}" ]; then
+            echo "::error::LLXPRT_DEFAULT_PROVIDER repository variable is required for real-model evals"
+            missing=1
+          fi
+          # Report which key variable was consulted (name only, never the value).
+          echo "Consulted API key secret named: ${KEY_VAR_NAME:-<unset>}"
```
[the remark this answers is not in the document]
Comment (acoliver) [reply .github/workflows/_evals-run.yml:81]: Retained the diagnostic intentionally. It prints only the configuration variable name, which is already visible in the repository workflow definition, never its secret value. Naming the missing variable makes preflight failures actionable without exposing credentials.

File: scripts/tests/aggregate-cli.test.js [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+      });
+      expect(result.exitCode, result.stderr).toBe(0);
+      expect(result.stdout).toContain('Eval Results Summary');
+    });
+  });
+
+  it('falls back to default [1,2,3] when env value is an empty string', () => {
+    // An empty AGGREGATE_EXPECTED_ATTEMPTS is the documented "not provided"
+    // behavior: it must NOT fail closed (unlike a malformed-but-nonempty
+    // value), but silently fall back to the default [1,2,3].
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/aggregate-cli.test.js]: Addressed. The comment now says that an empty value and must silently fall back, removing the ambiguous contrast.

File: scripts/aggregate-evals-historical.js [hunk 8]
```diff
@@ -0,0 +1,798 @@
...
+ * @type {RunSync}
+ */
+function spawnSyncBound(cmd, args) {
+  const result = spawnSync(cmd, args, { encoding: 'utf-8' });
+  return {
+    status: typeof result.status === 'number' ? result.status : 1,
+    stdout: typeof result.stdout === 'string' ? result.stdout : '',
+    stderr: typeof result.stderr === 'string' ? result.stderr : '',
+  };
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/aggregate-evals-historical.js:83]: Addressed. Workflow-run listing now has a 60-second subprocess timeout, preventing an indefinitely blocked synchronous GitHub CLI call.

File: scripts/aggregate-evals-historical.js [hunk 9]
```diff
@@ -0,0 +1,798 @@
...
+    {
+      encoding: 'utf-8',
+    },
+  );
+  return {
+    status: typeof result.status === 'number' ? result.status : 1,
+    stdout: typeof result.stdout === 'string' ? result.stdout : '',
+    stderr: typeof result.stderr === 'string' ? result.stderr : '',
+  };
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/aggregate-evals-historical.js:577]: Addressed. Artifact downloads now have a five-minute subprocess timeout, separately sized from the run-list request.

File: scripts/aggregate-evals-historical.js [hunk 10]
```diff
@@ -0,0 +1,798 @@
...
+ *   Tests inject a processor that throws for one run to prove loop-level
+ *   exception isolation: one bad run cannot abort the remaining runs.
+ * @returns {Map<string, Map<string, {pass: number, fail: number, total: number}>>}
+ *   Outer map: run ID -> inner map
+ *   Inner map: test name -> stats
+ */
+export function fetchHistoricalData(
+  listRunsPage = listRunsPageWithGh,
+  processRun = processHistoricalRun,
+) {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/aggregate-evals-historical.js:763]: Addressed. The JSDoc now documents the structured runs/rawCount/totalCount result and the retained bare-array compatibility form.

File: scripts/aggregate-evals-historical.js [hunk 11]
```diff
@@ -0,0 +1,798 @@
...
+  }
+  if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
+    console.error('Warning: workflow run envelope is not a JSON object');
+    return empty;
+  }
+  const src = /** @type {Record<string, unknown>} */ (parsed);
+  const runs = src['workflow_runs'];
+  if (!Array.isArray(runs)) {
+    return empty;
+  }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/aggregate-evals-historical.js:425]: Addressed. A missing or non-array workflow_runs field now emits an explicit warning, with behavioral coverage that captures and verifies the diagnostic.

File: scripts/aggregate-evals-historical.js [hunk 12]
```diff
@@ -0,0 +1,798 @@
...
+  const { runStats, reportErrors } = parseHistoricalReportsStrict(reports);
+  if (reportErrors.length > 0) {
+    const reason = `malformed report: ${reportErrors.join('; ')}`;
+    console.error(`Warning: Omitting historical run ${runId}: ${reason}`);
+    return { ...empty, reason };
+  }
+
+  if (runStats.size === 0) {
+    return { ...empty, reason: 'no usable assertions in any report' };
+  }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/aggregate-evals-historical.js:740]: Addressed. A successful artifact download containing no reports now returns the explicit reason no reports found in downloaded artifacts before cardinality or usability checks.

File: scripts/tests/evals-save-memory-assertion.test.js [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+    saveMemoryFactEquals = mod.saveMemoryFactEquals;
+  });
+
+  it('accepts the canonical fact exactly', () => {
+    const predicate = saveMemoryFactEquals;
+    const match = predicate('blue');
+    expect(match(JSON.stringify({ fact: 'My favorite color is blue' }))).toBe(
+      true,
+    );
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/evals-save-memory-assertion.test.js:40]: Addressed. The redundant predicate and assertion aliases were removed; tests now call the descriptive loaded functions directly while retaining only useful per-token matcher variables.

File: scripts/tests/evals-save-memory-assertion.test.js [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+
+  it('accepts the answer with harmless outer whitespace', () => {
+    const assert = assertFavoriteColorBlueOutput;
+    expect(() => assert('   $blue$   ')).not.toThrow();
+  });
+
+  it('rejects "red" (wrong color, even dollar-wrapped)', () => {
+    const assert = assertFavoriteColorBlueOutput;
+    expect(() => assert('$red$')).toThrow(/exact answer/i);
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/evals-save-memory-assertion.test.js:202]: Retained these assertions intentionally as diagnostic-contract tests. The messages distinguish missing or non-string output from present but semantically incorrect output, and the helper documentation defines those separate failure classes. Matching a stable semantic phrase protects that distinction without asserting the complete wording.

File: scripts/aggregate_evals.js [hunk 15]
```diff
@@ -62,175 +117,52 @@ function findReportsEntry(fullPath, entry, reports) {
...
   }
 
-  return historical;
+  return {
+    valid: errors.length === 0 && usableAssertions > 0,
+    stats,
+    errors,
+    usableAssertions,
+    reports,
+  };
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/aggregate_evals.js:164]: Addressed. The unused reports property was removed from aggregateReports so its return contract contains only validity, statistics, errors, and usable assertion count.

File: scripts/tests/evals-typescript-compliance.test.js [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+  return block;
+}
+
+/**
+ * Resolve the global ignores from the actual ESLint config object. Flat config
+ * allows a top-level `ignores` property (an object with an `ignores` string
+ * array) and/or per-block ignores. Returns every ignore glob declared anywhere
+ * in the config so callers can assert none matches evals/**.
+ */
+async function loadEslintIgnores() {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/evals-typescript-compliance.test.js:49]: Addressed. The comment now accurately states that the helper collects ignores arrays from every ESLint flat-config block.
